internals/plan/plan.go

```diff
@@ -1085,19 +1208,90 @@ func (p *Plan) checkCycles() error {
 }
 
 func ParseLayer(order int, label string, data []byte) (*Layer, error) {
-	layer := Layer{
-		Services:   map[string]*Service{},
-		Checks:     map[string]*Check{},
-		LogTargets: map[string]*LogTarget{},
-	}
-	dec := yaml.NewDecoder(bytes.NewBuffer(data))
-	dec.KnownFields(true)
-	err := dec.Decode(&layer)
+	layer := &Layer{
+		Services:   make(map[string]*Service),
+		Checks:     make(map[string]*Check),
+		LogTargets: make(map[string]*LogTarget),
+		Sections:   make(map[string]LayerSection),
+	}
+
+	// The following manual approach is required because:
+	//
+	// 1. Extended sections are YAML inlined, and also do not have a
+	// concrete type at this level, we cannot simply unmarshal the layer
+	// in one step.
+	//
+	// 2. We honor KnownFields = true behaviour for non extended schema
+	// sections, and at the top field level, which includes Section field
+	// names.
+	builtinSections := map[string]interface{}{
+		"summary":     &layer.Summary,
+		"description": &layer.Description,
+		"services":    &layer.Services,
+		"checks":      &layer.Checks,
+		"log-targets": &layer.LogTargets,
+	}
+	// Make sure builtinSections contains the exact same fields as expected
+	// in the Layer type.
+	if !mapHasKeys(builtinSections, layerBuiltins) {
```

Comment:
mapHasKeys doesn't sound like a bijective assertion - that is, typically hasKey is true if the key is present, not if it is the only key. That's how I read this function and was confused about how it satisfied the comment above.

Reply:
I changed this to ```mapMatchKeys```.

@benhoyt Do you have a better proposal ?

Reply:
Further feedback from Ben and further thinking, I've moved section field change detection purely into the test domain. Remove this runtime check, as at best it only covers some areas in the code, not all. Resolving.